### ISSUE TITLE
Error: Database type oci not supported.

### ISSUE THREAD

**Issue by ndunand** [ndunand/moodle-tool_mergeusers#152]
Oracle database type not supported.

This is because we don't check for 'oci' in https://github.com/ndunand/moodle-tool_mergeusers/blob/master/lib/mergeusertool.php#L140 .

**Comment by ndunand** [comment]
to try: run `SELECXT table_name FROM sys.all_tables WHERE table_name LIKLE 'mdl_%';` in Oracle when logged in as the Moodle user.

**Comment by hmgp** [comment]
Hi. Tried it on Oracle. There are some problems.
1 - Mistype I think, it's LIKE, you have LIKLE
2 - On Oracle we can't do mld_ has a prefix. In my case I have M_. Is there any place on Moodle where the Prefix used is available? If so, this could be used instead of a fixed string.
3 - In Oracle mdl is different than MDL, so we have to make extra to guarantee the select works, like put a lower() or upper() on the comparison. Example: SELECT table_name FROM sys.all_tables WHERE lower(table_name) LIKE 'mdl_%';

Thank you very much. If you need anything else just tell me.

**Comment by ndunand** [comment]
Hi,
1. Thanks, indeed it is `LIKE`, I should have double checked.
2. Yes, `mdl_` is default, but the actual value is stored in `$CFG->prefix`
3. point 2 above should take care of this

**Comment by hmgp** [comment]
You're totally wright, point 2 takes care of point 3.

**Comment by ndunand** [comment]
Do you have a test Moodle version (on Oracle) where you could test a modified version of the plugin? If so, you could add

```
            case 'oci':
                $this->sqlListTables = "SELECT table_name FROM sys.all_tables WHERE table_name LIKE '" .
                        $CFG->prefix . "%'";
                break;
```

on the line above `default` in https://github.com/ndunand/moodle-tool_mergeusers/blob/master/lib/mergeusertool.php#L154

**Comment by hmgp** [comment]
Of course :-) 
I will give feedback in a moment.

**Comment by hmgp** [comment]
With the change it works. Now I get the chance to merge users.
I will create some test cases in my test Moodle enviroment and I will report back.
This might take some time as we are in a transition of a 3.4 production enviroment to a 3.7.

Thnak you very much for this help!

**Comment by ndunand** [comment]
Thanks. As soon as you report back, and if we can confirm all is OK with Oracle, I will release a new version including this fix, so you won't have to redo the fix on each update.

**Comment by hmgp** [comment]
Hello. I don't know if we are doing something wrong or not but I don't think it is merging the users.
Our use case was:
2 user both are enrolled in a course with a SCORM activitie. 
- User 1 doesn't have any record on the course. Never entered the course, is only enrolled.
- User 2 submits the SCORM activitie and has score, status, location, suspend_data and time records. Also get's a grade for it.
- Merge User 2 into User 1.
- I get no errors, In the end says Merge successful. But nothing changed except User 2 is suspended (as it should).
Another thing I don't understand is you have a line which says: "Here are the queries that have been sent to the DB:" but there are none. I attach a log for your understanding

[https://photos.app.goo.gl/DSGPDoH45uLdrKNe9](url)

Are we missing something in this use case? Are we doing anything wrong?

Thank you.

P.S.: We are testing in a Moodle 3.7.1 which is what we are upgrading to in the next weeks.

**Comment by ndunand** [comment]
It looks like, as you noticed, no changes have been made to the database (except suspending User 2).

Do you get any more info if you set all debugging messages to be displayed?

**Comment by jpahullo** [comment]
Hi! 

The conditions regarding the line: https://github.com/ndunand/moodle-tool_mergeusers/blob/master/lib/mergeusertool.php#L140

are related to get the list of tables from the database.

Actually, Moodle overcomes this issue by this means:

```
global $DB;
$dbman = $DB->get_tables(false); //false to prevent using cache
```
So we don't need to reinvent the wheel, don't you think? Using this snippet, we can support this plugin with every database Moodle actually supports.

Let me know what do you think.

Thank you for your time and contribution!

**Comment by ndunand** [comment]
Hi @jpahullo !
Yes indeed you are right, it will make the code more readable and simple.

**Comment by hmgp** [comment]
Hello. I got this in debug developer:

line 316 of /admin/tool/mergeusers/lib/mergeusertool.php: call to MergeUserTool->updateGrades()
line 232 of /admin/tool/mergeusers/lib/mergeusertool.php: call to MergeUserTool->_merge()
line 135 of /admin/tool/mergeusers/index.php: call to MergeUserTool->merge()

Did you remember to make the first column something unique in your call to get_records? Duplicate value '2' found in column 'iteminstance'.
line 1194 of /lib/dml/oci_native_moodle_database.php: call to debugging()
line 523 of /admin/tool/mergeusers/lib/mergeusertool.php: call to oci_native_moodle_database->get_records_sql()
line 316 of /admin/tool/mergeusers/lib/mergeusertool.php: call to MergeUserTool->updateGrades()
line 232 of /admin/tool/mergeusers/lib/mergeusertool.php: call to MergeUserTool->_merge()
line 135 of /admin/tool/mergeusers/index.php: call to MergeUserTool->merge()

Did you remember to make the first column something unique in your call to get_records? Duplicate value '3' found in column 'iteminstance'.
line 1194 of /lib/dml/oci_native_moodle_database.php: call to debugging()
line 523 of /admin/tool/mergeusers/lib/mergeusertool.php: call to oci_native_moodle_database->get_records_sql()
line 316 of /admin/tool/mergeusers/lib/mergeusertool.php: call to MergeUserTool->updateGrades()
line 232 of /admin/tool/mergeusers/lib/mergeusertool.php: call to MergeUserTool->_merge()
line 135 of /admin/tool/mergeusers/index.php: call to MergeUserTool->merge()

I see by this it has to do with the grades, we have the option "Merge attempts from both user and renumber".

**Comment by jpahullo** [comment]
Hi!

We also see these warnings at our institution. We couldn't take a look at
it close.

Thanks for sharing!

**Comment by hmgp** [comment]
Hello. I did not understood what you mean. You wont look into this, is that it?

Thank you.

**Comment by jpahullo** [comment]
Hi!

Sorry for the confusion.

I mean that we also saw this warning while merging users. But that since now, I couldn't face this warning. And that, de facto, all attempts are merged properly even though seeing this warnings.

Above all, I think these warnings on quiz attempts need another issue. This is just for supporting the Oracle database. Could you create this new issue please?

Thanks for reporting and helping!

Jordi

**Comment by hmgp** [comment]
Ok, I already opened another issue. 

Thank you very much!

**Comment by jpahullo** [comment]
This is solved by PR #154
